src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs

```diff
@@ -119,6 +121,297 @@ public static JsonElement CreateJsonElement(string json)
 #endif
         }
 
+        internal static object CreateObjectFromJsonElement(JsonElement jsonElement)
+        {
+            if (jsonElement.ValueKind == JsonValueKind.String)
+            {
+                if (DateTime.TryParse(jsonElement.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out DateTime dateTime))
+                    return dateTime;
+
+                return jsonElement.GetString();
+            }
+            else if (jsonElement.ValueKind == JsonValueKind.False)
+                return false;
+            else if (jsonElement.ValueKind == JsonValueKind.True)
+                return true;
+            else if (jsonElement.ValueKind == JsonValueKind.Number)
+            {
+                if (jsonElement.TryGetInt32(out int intValue))
+                    return intValue;
+                else if (jsonElement.TryGetInt64(out long longValue))
+                    return longValue;
+                else if (jsonElement.TryGetDecimal(out decimal decimalValue))
+                    return decimalValue;
```

Comment:
nit: move this last as it's more specific and maybe less common compared to the other numeric types?